lib/collection.js

```diff
@@ -290,149 +284,9 @@ Collection.prototype.find = deprecateOptions(
-    // // If we have overridden slaveOk otherwise use the default db setting
-    newOptions.slaveOk = options.slaveOk != null ? options.slaveOk : this.s.db.slaveOk;
-
-    // Add read preference if needed
-    newOptions.readPreference = resolveReadPreference(newOptions, {
-      db: this.s.db,
-      collection: this
-    });
-
-    // Set slave ok to true if read preference different from primary
-    if (
-      newOptions.readPreference != null &&
-      (newOptions.readPreference !== 'primary' || newOptions.readPreference.mode !== 'primary')
-    ) {
-      newOptions.slaveOk = true;
-    }
-
-    // Ensure the query is an object
-    if (selector != null && typeof selector !== 'object') {
-      throw MongoError.create({ message: 'query selector must be an object', driver: true });
-    }
-
-    // Build the find command
-    const findCommand = {
-      find: this.s.namespace,
-      limit: newOptions.limit,
-      skip: newOptions.skip,
-      query: selector
-    };
-
-    // Ensure we use the right await data option
-    if (typeof newOptions.awaitdata === 'boolean') {
-      newOptions.awaitData = newOptions.awaitdata;
-    }
-
-    // Translate to new command option noCursorTimeout
-    if (typeof newOptions.timeout === 'boolean') newOptions.noCursorTimeout = newOptions.timeout;
-
-    decorateCommand(findCommand, newOptions, ['session', 'collation']);
-
-    if (projection) findCommand.fields = projection;
-
-    // Add db object to the new options
-    newOptions.db = this.s.db;
-
-    // Add the promise library
-    newOptions.promiseLibrary = this.s.promiseLibrary;
-
-    // Set raw if available at collection level
-    if (newOptions.raw == null && typeof this.s.raw === 'boolean') newOptions.raw = this.s.raw;
-    // Set promoteLongs if available at collection level
-    if (newOptions.promoteLongs == null && typeof this.s.promoteLongs === 'boolean')
-      newOptions.promoteLongs = this.s.promoteLongs;
-    if (newOptions.promoteValues == null && typeof this.s.promoteValues === 'boolean')
-      newOptions.promoteValues = this.s.promoteValues;
-    if (newOptions.promoteBuffers == null && typeof this.s.promoteBuffers === 'boolean')
-      newOptions.promoteBuffers = this.s.promoteBuffers;
-
-    // Sort options
-    if (findCommand.sort) {
-      findCommand.sort = formattedOrderClause(findCommand.sort);
-    }
-
-    // Set the readConcern
-    decorateWithReadConcern(findCommand, this, options);
-
-    // Decorate find command with collation options
-    try {
-      decorateWithCollation(findCommand, this, options);
-    } catch (err) {
-      if (typeof callback === 'function') return callback(err, null);
-      throw err;
-    }
-
-    const cursor = this.s.topology.cursor(this.s.namespace, findCommand, newOptions);
+    const findOperation = new FindOperation(this, query, options);
 
-    return typeof callback === 'function' ? handleCallback(callback, null, cursor) : cursor;
+    return findOperation.execute(callback);
```

Comment:
This makes me think that we should treat cursor-generation differently from operations. Find / Aggregate are both creating a cursor synchronously, and then that cursor executes the operations against the server. I don't know if these should be grouped with our Operations (for one, most of the traits are not going to apply).